client/src/JobSelector.jsx

```diff
@@ -1,56 +1,57 @@
 
-class JobSelector extends React.Component {
-    constructor(props) {
-        super(props);
-        this.state = { active: true, data: [] };
-        this.ajax = AjaxHelper.getInstance();
-    }
-
-    hide() {
-        var newState = this.state;
-        newState.active = false;
-        this.setState(newState);
-    }
-
-    show() {
-        this.setState({ active: true, data: [] });
-        this.componentDidMount();
-    }
-
-    componentDidMount() {
-        this.ajax.getJSON(
-            this.ajax.ingestApiUrl,
+const JobSelector = () => {
+    const [active, setActive] = useState(true);
+    const [data, setData] = useState([]);
+    const ajax = AjaxHelper.getInstance();
+
+    const hide = () => {
+        setActive(false);
+    };
+    const show = () => {
+        setActive(true);
+        ajax.getJSON(
+            ajax.ingestApiUrl,
             null,
-            function (data) {
-                this.setState({ active: true, data: data });
-            }.bind(this)
-        );
-    }
-
-    render() {
-        var categories = [];
-        var empty_categories = [];
-        for (var i in this.state.data) {
-            var category = this.state.data[i];
-            var element = <Category key={category.category} data={category} />;
-            if (category.jobs.length > 0) {
-                categories[categories.length] = element;
-            } else {
-                empty_categories[empty_categories.length] = element;
+            (data) => {
+                setActive(true);
+                setData(data);
             }
-        }
-        return (
-            <div className={this.state.active ? "" : "hidden"} id="jobSelector">
-                {categories}
-                {empty_categories}
-            </div>
         );
-    }
+    };
+
+    var categories = [];
+    var empty_categories = [];
+    const categoryComponents = Object.keys(data).reduce((categoryComponents, key) => {
+        const category = data[key];
+        const categoryComponent = <Category key={category.category} data={category} />;
+        if(category.jobs.length > 0) {
+            categoryComponents[0] = [...categoryComponents[0], categoryComponent];
+        } else {
+            categoryComponents[1] = [...categoryComponents[1], categoryComponent];
+        }
+        return categoryComponents;
+    },[[],[]]).reduce((a,b) => [...a, ...b],[]);
+
+    // for (var i in this.state.data) {
+    //     var category = this.state.data[i];
+    //     var element = <Category key={category.category} data={category} />;
+    //     if (category.jobs.length > 0) {
+    //         categories[categories.length] = element;
+    //     } else {
+    //         empty_categories[empty_categories.length] = element;
+    //     }
+    // }
+    return (
+        <div className={this.state.active ? "" : "hidden"} id="jobSelector">
+            {categoryComponents}
+        </div>
+    );
+
 }
 
 JobSelector.propTypes = {
```

Comment:
Do these propTypes definitions still mean anything after conversion to functional components? Do they need to be updated or removed? I think @crhallberg added them in the first place, so he may know more about their purpose if it is not already clear.

Reply:
The linter was complaining about them, so I kept them.

Reply:
I see... but I suspect they are no longer accurate. Something we should probably revisit with @crhallberg later; maybe worth putting a task on the board to review them at the end of the refactoring, especially if we're able to reduce/simplify.